pkg/controller/certificaterequests/venafi/venafi.go

```diff
@@ -104,31 +112,51 @@ func (v *Venafi) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerO
 		}
 	}
 
-	certPem, err := client.Sign(cr.Spec.CSRPEM, duration, customFields)
+	pickupID, exists := cr.ObjectMeta.Annotations[VenafiPickupIDAnnotation]
 
-	// Check some known error types
-	if err != nil {
-		switch err.(type) {
+	// check if the pickup ID annotation is there, if not set it up.
+	if !exists || pickupID == "" {
+		pickupID, err = client.RequestCertificate(cr.Spec.CSRPEM, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
 
-		case venafiinternal.ErrCustomFieldsType:
-			v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
-			log.Error(err, err.Error())
+			case venaficlient.ErrCustomFieldsType:
+				v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
+				log.Error(err, err.Error())
 
-			return nil, nil
+				return nil, nil
 
-		case endpoint.ErrCertificatePending:
-			message := "Venafi certificate still in a pending state, the request will be retried"
+			default:
+				message := "Failed to request venafi certificate"
 
-			v.reporter.Pending(cr, err, "IssuancePending", message)
-			log.Error(err, message)
+				v.reporter.Failed(cr, err, "RequestError", message)
+				log.Error(err, message)
+
+				return nil, err
+			}
+		}
+
+		v.reporter.Pending(cr, err, "IssuancePending", "Venafi certificate is requested")
+
+		metav1.SetMetaDataAnnotation(&cr.ObjectMeta, VenafiPickupIDAnnotation, pickupID)
+		_, err = v.cmClient.CertmanagerV1alpha2().CertificateRequests(cr.GetNamespace()).Update(ctx, cr, metav1.UpdateOptions{})
+		if err != nil {
 			return nil, err
+		}
+
+		return nil, nil
+	}
 
-		case endpoint.ErrRetrieveCertificateTimeout:
-			message := "Timed out waiting for venafi certificate, the request will be retried"
+	certPem, err := client.RetreiveCertificate(pickupID, cr.Spec.CSRPEM, duration, customFields)
+	if err != nil {
+		switch err.(type) {
+		case endpoint.ErrCertificatePending, endpoint.ErrRetrieveCertificateTimeout:
+			message := "Venafi certificate still in a pending state, the request will be retried"
 
-			v.reporter.Failed(cr, err, "Timeout", message)
+			v.reporter.Pending(cr, err, "IssuancePending", message)
 			log.Error(err, message)
-			return nil, nil
+			return nil, err
```

Comment:
We took this approach when we first made the ACME issuer async (returning errors to trigger retries). It got _complicated_, and messy. Happy to keep it as-is for now, but we _really_ should fix this up. My comment above around removing the 'Sign' method concept will also really help here, as it'll save the layer of indirection that forces us to use errors for this behaviour.

Reply:
Deferring this for #3074